src/test/java/com/redhat/foreman/ForemanSharedNodeCloudTest.java

```diff
@@ -243,4 +251,69 @@ public void testRoundTrip() throws IOException, URISyntaxException, InterruptedE
         assertTrue(initialComputerSet == finalComputerSet);
     }
 
+    /**
+     * Round trip test that simulates a loss of connection to
+     * Foreman.
+     * @throws IOException if occurs.
+     * @throws URISyntaxException if occurs.
+     * @throws InterruptedException if occurs.
+     */
+    @Test
+    public void testWithLossOfConnection() throws Exception {
+        final CountDownLatch disposeCheckLatch = new CountDownLatch(60);
+        final CountDownLatch cleanedCheckLatch = new CountDownLatch(60);
+
+        setupWireMock();
+        // Add cloud
+        ForemanSharedNodeCloud fCloud = new ForemanSharedNodeCloud("mycloud", URL,
+                USER, Secret.fromString(PASSWORD), "", 1);
+
+        fCloud.setLauncherFactory(new ForemanDummyComputerLauncherFactory());
+        j.getInstance().clouds.add(fCloud);
+
+        final OneShotEvent buildStarted = new OneShotEvent();
+        FreeStyleProject job = j.createFreeStyleProject();
+        job.setAssignedLabel(new LabelAtom("label1"));
+        job.getBuildersList().add(new TestBuilder() {
+            public boolean perform(AbstractBuild<?, ?> build, Launcher launcher,
+                                   BuildListener listener) throws InterruptedException, IOException {
+                buildStarted.signal();
+                return true;
+            }
+        });
+
+        assertTrue(job.scheduleBuild(0, new UserIdCause()));
+        buildStarted.block();
+        // Let's simulate a Foreman connection error
+        stubFor(get(urlMatching("/api/.*"))
+                .willReturn(aResponse().withStatus(HTTPERROR)));
+        TestUtils.waitForBuilds(job, 1);
```

Comment:
Exactly. Reordering isn't possible because we need to break a connection through redefining via `stubFor(...)` right after job has started, *but* before it finishes.